wagtail/admin/views/collections.py

```diff
@@ -1,50 +1,138 @@
 
 
 class Index(IndexView):
-    permission_policy = collection_permission_policy
     model = Collection
     context_object_name = 'collections'
     template_name = 'wagtailadmin/collections/index.html'
-    add_url_name = 'wagtailadmin_collections:add'
-    page_title = ugettext_lazy("Collections")
-    add_item_label = ugettext_lazy("Add a collection")
     header_icon = 'folder-open-1'
 
+    def __init__(self):
+        super(Index, self).__init__()
+        self.parent_collection = None
+        self.is_searching = False
+
+    @property
+    def page_title(self):
+        if not self.parent_collection:
+            return ugettext_lazy("Collections")
+        return ugettext_lazy(self.parent_collection.name)
+
+    def get_paginate_by(self, queryset):
+        return 50
+
     def get_queryset(self):
-        # Only return children of the root node, so that the root is not editable
-        return Collection.get_first_root_node().get_children()
+        search = SearchForm(self.request.GET)
+
+        if search.is_valid() and search.cleaned_data['q']:
+            self.is_searching = True
+            return Collection.objects.filter(name__icontains=search.cleaned_data['q'])
+        else:
+            if not self.parent_collection:
+                # Find the root collection that the user has access to
+                self.parent_collection = get_explorable_root_collection(self.request.user)
 
+            return self.parent_collection.get_children()
 
-class Create(CreateView):
-    permission_policy = collection_permission_policy
+    def get_context_data(self):
+        if self.args:
+            parent_collection_id = self.args[0]
+            # TODO: Make sure user cannot get to a collection if they have the URL but not the perms
+            self.parent_collection = get_object_or_404(Collection, pk=parent_collection_id)
+
+        context = super(Index, self).get_context_data(object_list=self.get_queryset())
+
+        context.update({
+            'search_form': SearchForm(),
+            'parent_collection': self.parent_collection,
+            'parent_perms': self.parent_collection.permissions_for_user(self.request.user),
+        })
+
+        return context
+
+
+class Create(CollectionPermissionMixin, CreateView):
     form_class = CollectionForm
     page_title = ugettext_lazy("Add collection")
     success_message = ugettext_lazy("Collection '{0}' created.")
-    add_url_name = 'wagtailadmin_collections:add'
+    add_url_name = 'wagtailadmin_collections:add_child'
     edit_url_name = 'wagtailadmin_collections:edit'
-    index_url_name = 'wagtailadmin_collections:index'
     header_icon = 'folder-open-1'
 
+    def __init__(self):
+        super(Create, self).__init__()
+        self._parent_collection = None
+
+    def get_success_url(self):
+        return reverse('wagtailadmin_collections:parent_index', args=(self.parent_collection.pk,))
+
+    @property
+    def parent_collection(self):
```

Comment:
The numerous methods around `parent_collection` here and the somewhat duplication of functionality with the `Index` view above makes me wonder if there's a simpler way to do this. I'd like to look at this a little further but don't want to hold back the rest of the PR review.